### PR TITLE
[DNM] avoid infinite sync loop to consume bandwidth for nothing

### PR DESCRIPTION
**Pull request by garyyu**
A small improvement on https://github.com/mimblewimble/grin/issues/1684#issuecomment-429494942

In case a node is stuck for some reason, we will enter an infinite loop to download same Headers or same block from peers, this infinite loop which consume the Grin network bandwidth for nothing.

This PR will detect such kind of sync stuck and quit sync thread.

### CODE REVIEW

File: servers/src/grin/sync/body_sync.rs [hunk 1]
```diff
@@ -142,6 +144,17 @@ impl BodySync {
 			.collect::<Vec<_>>();
 
 		if hashes_to_get.len() > 0 {
+			// Stuck detection:
+			// In case stuck detected, exit sync thread to avoid useless infinite loop.
+			if hashes_to_get.len() == 1 && hashes_to_get[0].clone() == self.stuck_detector.0 {
+				self.stuck_detector.1 += 1;
+				if self.stuck_detector.1 >= 100 {
+					panic!("sync: body sync stuck detected! sync thread exited!");
```
Comment (sesam) [inline servers/src/grin/sync/body_sync.rs]: sorry I'm out of the loop, what does panic! do? Kill the current thread? Will it get resurrected, and if yes, where would that thread usually be killed if no stuck detected?

Comment (garyyu) [reply servers/src/grin/sync/body_sync.rs]: `panic!` here is used here to exit the sync thread.
no, this thread will not get resurrected.

**Comment by garyyu** [comment]
> partial code duplication with magic numbers (100 and +1).
No, it's not `duplication`. One is for HeaderSync, and another one is for BodySync.

**Comment by ignopeverell** [comment]
I empathize with the intent but I think this will mask 2 issues that need fixing (and so shouldn't be masked). Namely:

1. A node can get stuck.
2. Other peers don't do a great job at protecting themselves against abusive or malfunctioning peers.

**Comment by antiochp** [comment]
I think I agree with @ignopeverell on (2) here - this PR means a node will decide to kill its own sync process based on data from a connected (and potentially misbehaving, intentional or otherwise) peer.

* Say I'm a peer and I advertise a fork/chain with more work. 
* But I never actually give you those new headers and instead just keep giving you the same set of headers repeatedly.
* You notice this and eventually kill your own sync process.
* All the other nodes on the network do the same thing.

We introduce a vector for a network wide DDoS attack with this.

It's really hard to know if we are actually "stuck" ourselves or if we appear to be "stuck" because of data from another peer.
And I think we need to assume the worst (or at least be prepared to handle the worst) in each situation.

**Comment by ignopeverell** [comment]
Agree with @antiochp too, it seems gameable. And what's good with having a lagging node? We might as well shutdown the node at this point.

**Comment by garyyu** [comment]
@antiochp  thanks for your review~

> this PR means a node will decide to kill its own sync process based on data from a ... peer.

Pls note, not based on **a** peer, I use **100** that should cover all connected peers.
But I do agree I need check whether the connected peers number is just **ONE** or less than a threshold. Will give improvement on this.

**Comment by garyyu** [comment]
@ignopeverell  thanks for your review~

> Agree with @antiochp too, it seems gameable. And what's good with having a lagging node? We might as well shutdown the node at this point.

Sorry, I'm confused by this comment. 
- When you said "`might shutdown the node at this point`", that means you second this PR. 
- When you said "`Agree with @antiochp `", but @antiochp is obvious against this PR because it looks like this PR introduce a network wide DDoS attack risk (which I can improve to avoid) 😄

**Comment by antiochp** [comment]
> Pls note, not based on a peer, I use 100 that should cover all connected peers.

But this is not how the network works. We sync headers from the "most work peer" which may be a single malicious peer or a group of several malicious peers, either way the result is the same.

We need to verify that a peer is actually being honest when it advertises a fork with more work. 
The _only_ way we can do that is by downloading all the headers on that fork to validate the PoW.

We have not control over what other nodes tell us.
Peers can be lying to us, either unintentionally or maliciously.
All we can do is decide to do is ignore them and/or ban them.

My point was that we cannot make decisions locally around shutting our node down (or part of the node, like the sync thread in this case) based on what other (untrusted) nodes on the network tell us.

I don't think this is the right approach to solving this problem.

**Comment by garyyu** [comment]
@antiochp 
> My point was that we cannot make decisions locally around shutting our node down (or part of the node, like the sync thread in this case) based on what other (untrusted) nodes on the network tell us.

Agree. Let me think it again.

**Pull request closed without merging**